File: scripts/laser2pc.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import PointCloud2
from sensor_msgs.msg import LaserScan
from laser_geometry import LaserProjection


class Laser2PC():
    def __init__(self):
        self.laserProj = LaserProjection()
        self.pcPub = rospy.Publisher("/cloud", PointCloud2, queue_size=100)
        self.laserSub = rospy.Subscriber(
            "/scan", LaserScan, self.laserCallback)

    def laserCallback(self, data):
        out = self.laserProj.projectLaser(data)
        self.pcPub.publish(out)


if __name__ == "__main__":
    rospy.init_node("laser2PC")
    node = Laser2PC()
    rospy.spin()

File: src/obsolete/laser2pc.cpp
#include <ros/ros.h>
#include <tf/transform_listener.h>
#include <sensor_msgs/LaserScan.h>
#include <laser_geometry/laser_geometry.h>

class Laser2PC
{
public:
    Laser2PC();
    void scanCallback(const sensor_msgs::LaserScan::ConstPtr &scan);

private:
    ros::NodeHandle node_;
    laser_geometry::LaserProjection projector_;
    tf::TransformListener tfListener_;
    ros::Publisher pcPub_;
    ros::Subscriber laserSub_;
};

Laser2PC::Laser2PC()
{
    laserSub_ = node_.subscribe<sensor_msgs::LaserScan>("/scan", 100, &Laser2PC::scanCallback, this);
    pcPub_ = node_.advertise<sensor_msgs::PointCloud2>("/cloud", 100, false);
}

void Laser2PC::scanCallback(const sensor_msgs::LaserScan::ConstPtr &scan)
{
    if (!tfListener_.waitForTransform(
            scan->header.frame_id,
            "/base_link",
            scan->header.stamp + ros::Duration().fromSec(scan->ranges.size() * scan->time_increment), ros::Duration(1.0)))
    {
        return;
    }

    sensor_msgs::PointCloud2 cloud;
    projector_.transformLaserScanToPointCloud("base_link", *scan, cloud, tfListener_);
    pcPub_.publish(cloud);
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "Laser2PC");
    Laser2PC converter;
    ros::spin();
    return 0;
}